credentials/xds/xds.go

```diff
@@ -0,0 +1,332 @@
+}
+
+// NewClientCredentials returns a new client-side transport credentials
+// implementation which uses xDS APIs to fetch its security configuration.
+func NewClientCredentials(opts ClientOptions) (credentials.TransportCredentials, error) {
+	if opts.FallbackCreds == nil {
+		return nil, errors.New("missing fallback credentials")
+	}
+	return &credsImpl{
+		isClient: true,
+		fallback: opts.FallbackCreds,
+	}, nil
+}
+
+// credsImpl is an implementation of the credentials.TransportCredentials
+// interface which uses xDS APIs to fetch its security configuration.
+type credsImpl struct {
+	isClient bool
+	fallback credentials.TransportCredentials
+}
+
+// handshakeCtxKey is the context key used to store HandshakeInfo values.
+type handshakeCtxKey struct{}
+
+// HandshakeInfo wraps all the security configuration required by client and
+// server handshake methods in credsImpl. The xDS implementation will be
+// responsible for populating these fields.
+//
+// Safe for concurrent access.
+type HandshakeInfo struct {
+	mu               sync.Mutex
+	rootProvider     certprovider.Provider
+	identityProvider certprovider.Provider
+	acceptedSANs     []string // Only on the client side.
+}
+
+// SetRootCertProvider updates the root certificate provider.
+func (chi *HandshakeInfo) SetRootCertProvider(root certprovider.Provider) {
+	chi.mu.Lock()
+	chi.rootProvider = root
+	chi.mu.Unlock()
+}
+
+// SetIdentityCertProvider updates the identity certificate provider.
+func (chi *HandshakeInfo) SetIdentityCertProvider(identity certprovider.Provider) {
+	chi.mu.Lock()
+	chi.identityProvider = identity
+	chi.mu.Unlock()
+}
+
+// SetAcceptedSANs updates the list of accepted SANs.
+func (chi *HandshakeInfo) SetAcceptedSANs(sans []string) {
+	chi.mu.Lock()
+	chi.acceptedSANs = sans
+	chi.mu.Unlock()
+}
+
+func (chi *HandshakeInfo) validate(isClient bool) error {
+	chi.mu.Lock()
+	defer chi.mu.Unlock()
+
+	// On the client side, rootProvider is mandatory. IdentityProvider is
+	// optional based on whether the client is doing TLS or mTLS.
+	if isClient && chi.rootProvider == nil {
+		return errors.New("root certificate provider is missing")
+	}
+
+	// On the server side, identityProvider is mandatory. RootProvider is
+	// optional based on whether the server is doing TLS or mTLS.
+	if !isClient && chi.identityProvider == nil {
+		return errors.New("identity certificate provider is missing")
+	}
+
+	return nil
+}
+
+func (chi *HandshakeInfo) matchingSANExists(cert *x509.Certificate) bool {
+	var sans []string
+	// SANs can be specified in any of these four fields on the parsed cert.
+	sans = append(sans, cert.DNSNames...)
+	sans = append(sans, cert.EmailAddresses...)
+	for _, ip := range cert.IPAddresses {
+		sans = append(sans, ip.String())
+	}
+	for _, uri := range cert.URIs {
+		sans = append(sans, uri.String())
+	}
+
+	chi.mu.Lock()
+	defer chi.mu.Unlock()
+	for _, san := range sans {
+		for _, asan := range chi.acceptedSANs {
```

Comment:
Currently we are using two for loops. Would it be more efficient if we use a set-like data structure for `sans`?

Reply:
Done. Thanks.